api.js

```diff
@@ -426,6 +427,84 @@ API.prototype.noticeError = function noticeError(error, customAttributes) {
   this.agent.errors.addUserError(transaction, error, filteredAttributes)
 }
 
+/**
+ * Sends an application log message to New Relic. The agent already
+ * automatically does this for some instrumented logging libraries,
+ * but in case you are using another logging method that is not
+ * already instrumented by the agent, you can use this function
+ * instead.
+ *
+ * If application log forwarding is disabled in the agent
+ * configuration, this function does nothing.
+ *
+ * @param {object} logEvent The log event object to send. Any
+ *   attributes besides `message`, `level`, `timestamp`, and `error` are
+ *   recorded unchanged.
+ * @param {string} logEvent.message The log message.
+ * @param {string} logEvent.level The log level severity. If this key is
+ *   missing, it will default to UNKNOWN
+ * @param {number} logEvent.timestamp   ECMAScript epoch number denoting the
+ *   time that this log message was produced. If this key is missing,
+ *   it will default to the output of `Date.now()`.
+ * @param {object} logEvent.error Object containing error details. Ignored if missing.
+ * @param {string} logEvent.error.message  The error message.
+ * @param {string} logEvent.error.stack The stack trace of the error.
+ * @param {string} logEvent.error.class The type of the error. If this key is missing,
+ *   defaults to 'Error'.
+ */
+API.prototype.recordLogEvent = function recordLogEvent(logEvent = {}) {
+  const metric = this.agent.metrics.getOrCreateMetric(NAMES.SUPPORTABILITY.API + '/recordLogEvent')
+  metric.incrementCallCount()
+
+  if (!applicationLogging.isLogForwardingEnabled(this.agent.config, this.agent)) {
+    logger.warnOnce(
+      'Record logs',
+      'Application log forwarding disabled, method API#recordLogEvent will not record messages'
+    )
+    return
+  }
+
+  // If they don't pass a logEvent object, or it doesn't have the
+  // required `message` key, bail out.
+  if (typeof logEvent !== 'object' || logEvent.message === undefined) {
+    logger.warn(
+      'recordLogEvent requires an object with a `message` attribute for its single argument, got %s (%s)',
+      stringify(logEvent),
+      typeof logEvent
+    )
+    return
+  }
+  logEvent.message = applicationLogging.truncate(logEvent.message)
+
+  if (!logEvent.level) {
+    logger.debug('no log level set, setting it to UNKNOWN')
+    logEvent.level = 'UNKNOWN'
+  }
+
+  if (typeof logEvent.timestamp !== 'number') {
+    logger.debug('no timestamp set, setting it to `Date.now()`')
+    logEvent.timestamp = Date.now()
+  }
+
+  if (logEvent.error) {
+    logEvent['error.message'] = applicationLogging.truncate(logEvent.error.message)
+    logEvent['error.stack'] = applicationLogging.truncate(logEvent.error.stack)
+    if (!logEvent.error.class) {
+      logEvent['error.class'] = 'Error'
```

Comment:
We don't want to assign this. we want to use whatever is on the logEvent.error.type

Reply:
Even if it's `undefined`? Do we just not set anything in that case?

Reply:
if this is truly an error it'll always be defined

Reply:
that's why this generic logging method is getting too heavy handed. if a custom included an error object in a log line that's not on an error key this logic won't get hit. i'm now leaning towards just ripping this out and seeing how our customers do this

Reply:
You mean `x.name`?

Reply:
I don't understand what you're asking

Reply:
```> x = new Error().type
undefined
> x = new Error().name
'Error'```

Reply:
ah crap, yes.  you can see our logic for winston is here

```js
  logLine['error.class'] =
    logLine.error.name === 'Error' ? logLine.error.constructor.name : logLine.error.name
```

Reply:
Okay, made the change, I think I like this version of the code now.